### DISCUSSION TITLE
Bench only one 1 cpu per time

### DISCUSSION THREAD

**Discussion by ducanh2110** [emqx/MQTTX#1439] [Q&A]
I am doing bench with mqttx but when I use top command to see only 1 CPU usage. Not all of my CPU. Anw to raise all cpu being used to make more connection?

**Comment by ysfscream** [comment]
Hi @ducanh2110, the MQTTX CLI is built on Node.js, which is inherently single-threaded, so by default, it won't utilize all CPU cores. We've recognized this limitation and plan to investigate ways to improve connection performance and count in future updates. 

Meanwhile, could you share more details about your current use case with the bench and the maximum connections you're achieving? It will help us benchmark and see where improvements can be made—apologies for the delayed response, and thanks for your feedback.

**Comment by ducanh2110** [comment]
Hi @ysfscream . We are aiming for create more than 15k ccu by the bench. We are using emqtt-bench but it bench with wss. We want to bench with wss

**Reply by ysfscream** [reply]
Cool, Have you made a comparison with emqtt-bench? We've tested mqttx cli on an Ubuntu machine with 2core 8G of RAM and achieved up to 60K connections. Are you encountering issues reaching beyond this with the mqttx bench now?

**Reply by ducanh2110** [reply]
yes. I've made the bench on the same VM. with emqtt-bench. They can use all of my core. Then they can make up to 15k connection. My VM is 16CPU, 16RAM running on Rhel8. I test with 9000 connections using mqttx-linux-64. But when it reach 9xx/9000. It become slowly then reach more than 1k maybe 10-50 conncetions. Then it keep there. I monitor the CPU usage it reach 100% core 1, their either do nothing

**Reply by ysfscream** [reply]
It seems that there are some differences between the MQTTX CLI Bench and emqtt-bench. @Red-Asuka We will optimize this issue in future versions.

**Comment by Red-Asuka** [comment]
Hi @ducanh2110 . I spent some time implementing the ability to enable multi-processes in the `bench conn` command and tested it on a Linux server with 4 cores and 24GB of RAM. By setting `sysctl -w net.ipv4.ip_local_port_range='1024 65535'`, even using only one core, I was able to achieve over 60k connections and fill up all available ports. The performance bottleneck is not on the CPU.

Unfortunately, it seems that there is no good way in Node.js to limit CPU usage to simulate reaching the bottleneck of a single process. This may require system-level restrictions, but unfortunately, I don't have a suitable testing environment at hand to test it. If you are interested, I can share with you an installation package for testing with this feature. I hope you can share your test results as well.
